### ISSUE TITLE
Bug Report:  Possible Payara JSF library bug / FISH-5675

### ISSUE THREAD

**Issue by rayquinlan** [payara/Payara#5346]
# Description #
I am running payara 5.202 with a Java/Primefaces based application. Everything works fine with PF8 but after upgrading to PF10 a particular error is occurring. This error appears to be thrown from the jakarta.faces.jar packaged with payara. More details on the bug are found here:

https://github.com/eclipse-ee4j/mojarra/issues/4936

Payara seem to have their own implementation in use when i check the manifest file of the above library, i see 
Bundle-Name: Mojarra JSF Implementation 2.3.14.payara-p2
Extension-Name: javax.faces
Bundle-Description: Eclipse Faces Implementation (jakarta.faces/2.3) 2
 .3.14.payara-p2
Implementation-Title: Mojarra
Implementation-Version: 2.3.14.payara-p2

It affects the <p:download and <p:dataExporter tags in PF10, causing a crash to occur.

## Expected Outcome ##
No crash

## Current Outcome ##
crash with following details
Stack Trace:
java.lang.ArrayIndexOutOfBoundsException: 17
	at javax.faces.component.UIComponentBase.restoreChildState(UIComponentBase.java:3418)
	at javax.faces.component.UIComponentBase.processRestoreState(UIComponentBase.java:1026)
	at javax.faces.component.UIComponentBase.restoreFacetsState(UIComponentBase.java:3445)
	at javax.faces.component.UIComponentBase.processRestoreState(UIComponentBase.java:1029)
	at javax.faces.component.UIViewRoot.processRestoreState(UIViewRoot.java:932)
	at com.sun.faces.application.view.JspStateManagementStrategy.restoreView(JspStateManagementStrategy.java:252)
	at com.sun.faces.application.StateManagerImpl.restoreView(StateManagerImpl.java:113)
	at com.sun.faces.application.view.ViewHandlingStrategy.restoreView(ViewHandlingStrategy.java:99)
	at com.sun.faces.application.view.FaceletViewHandlingStrategy.restoreView(FaceletViewHandlingStrategy.java:272)
	at com.sun.faces.application.view.MultiViewHandler.restoreView(MultiViewHandler.java:133)
	at javax.faces.application.ViewHandlerWrapper.restoreView(ViewHandlerWrapper.java:101)
	at com.sun.faces.lifecycle.RestoreViewPhase.execute(RestoreViewPhase.java:181)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:76)
	at com.sun.faces.lifecycle.RestoreViewPhase.doPhase(RestoreViewPhase.java:110)
	at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:177)
	at javax.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:707)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:451)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1636)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:331)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at com.huwag.hwoffice.layout.TrackingFilter.doFilter(TrackingFilter.java:114)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:253)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at com.huwag.hwoffice.log.util.HwLogFilter.doFilter(HwLogFilter.java:28)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:253)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:61)
	at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
	at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
	at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:450)
	at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
	at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
	at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
	at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:387)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:253)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at com.huwag.hwoffice.auth.keycloak.CustomKeycloakFilter.doFilter(CustomKeycloakFilter.java:149)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:253)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:257)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:161)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:757)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:577)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:158)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:520)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:217)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:182)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:156)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:218)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:95)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:524)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:89)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:94)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:33)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:114)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
	at java.lang.Thread.run(Thread.java:748)

[error-report.txt](https://github.com/payara/Payara/files/6816675/error-report.txt)

## Steps to reproduce ## 
See related raised issue above for details on it. 

## Environment ##

os.name=Windows 8.1
java.vm.specification.vendor=Oracle Corporation
java.version=1.8.0_152
glassfish.version=Payara Server  5.2021.1 #badassfish (build 2818)

**Comment by AlanRoth** [comment]
Hi @rayquinlan,

Can you please test on the latest release 5.2021.5, and if you can, provide a sample reproducer app for us to use on the latest version.

Thank you,
Alan

**Comment by rayquinlan** [comment]
Not sure why this would be fixed as there is no change in the library in question? I downloaded the latest version and looked at the jakartafaces.jar manifest. Its the same as the previous versions?

Implementation-Title: Mojarra
Implementation-Version: 2.3.14.payara-p2
Bundle-ManifestVersion: 2
Bundle-Vendor: Eclipse Foundation

I am unlikely to be able to repo this for you anytime soon(project commitments and on holidays), the issue has already been reproduced and accepted as a bug by the mojarra project themselves and a fix applied.  So payara should be looking to include a newer version of the jakarta-faces.jar file in their packaging to keep pace with their bug fixes?

Mojarra: 2.3.16 was the version of mojarra from the related issue, payara is only running on 2.3.14 and their own changes too by the looks of it. I already took a mojarra jar with the fixed attempt in it and deployed to payara instead of the packaged version but this caused further errors in the application.

**Comment by AlanRoth** [comment]
Hi @rayquinlan,

Its just policy to make sure bug reports are done on the latest version to rule out any variables.

If you are not able to provide a reproducer in the coming weeks, I can close this issue and you can open it when you are ready with a reproducer.

Thank you,
Alan

**Comment by rayquinlan** [comment]
Ok, so you are basically going to ignore a major bug in your software with
a major JSF provider? You have the bug reproduced in the related issue.  If
ye cant be bothered to investigate then perhaps I'll consider other app
servers.

**Comment by phillipross** [comment]
@AlanRoth If possible, can you keep this issue open for another week or so?

I'm preparing to migrate from PF8 to PF11 for some projects, and I'll see if I can reproduce the error on v5.2021.5.

Additionally, I looked at the upstream fixes for the mojarra bug and they appear to be very simple.  I'm not sure how the payara forks of mojarra are managed lifecycle-wise, but I'll also check to see if backporting the bugfixes to the payara fork is as straightforward as it looks 😉 

Thanks!

**Comment by melloware** [comment]
The fixes in Mojarra were quite simple in that it wasn't defensively checking the array before asking for the next value causing IndexOutOfBounds.

**Comment by folker-kuhn** [comment]
This should be a trivial update. Why is Payara using their own patched versions? Are those patches not provided upstream? Or is it just an early build of master from before 2.3.15 was released? If so, update to 2.3.16 should be really easy.

It looks like Payara has a big shortage of engineers atm, so even simple things are now difficult. Too bad.

**Comment by lprimak** [comment]
> It looks like Payara has a big shortage of engineers atm, so even simple things are now difficult. Too bad

that’s really unfair. Please remember that every little update needs to pass Jakarta TCK among other things. Engineering resources are funneled to customer priorities first. There a lot of things that may be a priority to some people that cannot be dealt with immediately.

**Comment by phillipross** [comment]
They've had to fork and patch many projects for awhile now (years) which predates their current community/enterprise edition and release cadence.  Mojarra specifically in the current v5 codebase is based on 2.3.14 (I think) but the v6 codebase uses mojarra 3.x 😉

**Comment by folker-kuhn** [comment]
> that’s really unfair.

Lenny, I really fail to see why this is unfair. Updating components used to be a routine business of Payara. This is even how Payara started! A fork of GlassFish that kept the components up to date.

Now we have to provide proof and reproducers and vote for component updates (https://github.com/payara/Payara/issues/5290) and test on the latest version even though the component clearly hasn't been updated? We're not talking about some new custom feature to be developed, but about simple component updates. The TCK for Mojarra has already been run by the Mojarra team. It won't give different results in Payara, but that's really besides the point.

This is what happened with Geronimo, Resin and Orion. Many engineers leaving, then the simplest of requests suddenly being difficult, then the product eventually halting.

We've been supporting Payara for years. please don't let us down.

**Comment by phillipross** [comment]
@AlanRoth I've verified that the issue does occur with 5.2021.5 using a slightly modified version of the reproducer that @melloware provided in the eclipse-ee4j/mojarra#4936 issue.  The modifications that were required were just removing the hardcoded geronimo/webbeans assumptions, but I'll zip up the modified reproducer and attach it to this issue.  Based on this verification, I think it'd be a good idea to keep this issue open, or at least create a separate issue to keep track of it. Thanks!

**Comment by phillipross** [comment]
Only tested against payara-server-full v5.2021.5:  [pf-7343.payara.zip](https://github.com/payara/Payara/files/6919023/pf-7343.payara.zip)

**Comment by smillidge** [comment]
@phillipross thanks for the test it's appreciated. 

@folker-kuhn the sock puppet returns, feel free to make a backport PR to our patched repo or a forward port of a Mojarra upgrade to master as you know so much about it and care so much

As this has been open less than a few weeks I think foretelling the demise of Payara is a little premature. 

If it is a simple fix we will get it done. @AlanRoth please create a jira.

**Comment by phillipross** [comment]
@smillidge i'm working on the backport now 😉 

@melloware what do you think the chances are of primefaces folks accepting a PR into the primefaces-test repo which uses maven profiles (or similar mechanisms) to ease the process of formulating and running tests cases within payara?

**Comment by melloware** [comment]
@phillipross Good chance it would be accepted as Thomas has wanted a non Jetty EE running profile for PrimeFaces Test for a while now.

**Comment by folker-kuhn** [comment]
> The issue is actually an example of why it's not always so simple and straightforward to update the components. Apparently they updated Yasson, it broke stuff, and they had to revert it?

I don't disagree, but instead of reverting it, fix whatever stuff it broke?

@smillidge I said it before, I'm passionate about Payara (just as I was about Geronimo back then). The Geronimo committers may remember me fondly for pushing for several fixes and accepting my pull requests, but as everyone was leaving there it was a useless exercise.

It's imho problematic that so many engineers are leaving Payara at the same time. It's not about this specific issue, that would be nonsense. It's that this issue is another data point in a pattern. If I look at your commit graph, all the top engineers have left. Some of them I see working now at projects that are very similar to Payara. The commits done to Payara have plummeted, meaning you either are working much more on closed source, or you have a shortage of engineers. Both don't give me much confidence.

You may not see it, but I wish Payara the best, and hope that whatever is causing all those engineers to leave is addressed.

**Comment by phillipross** [comment]
I've backported the mojarra fixes to the patched source payara fork and submitted a PR for that as can be seen in this issue.  I also forked the payara repo, created a branch from the 5.2021.5 tag, modified the build to use the patched mojarra, built the payara-server-full distro, and verified the reproducer now runs successfully with the patched payara.  I've also verified that the master branch builds cleanly against the patched mojarra as well.  Thanks to everyone who assisted!
